### PR TITLE
Add destination cluster info to response cookie

### PR DESCRIPTION
**Pull request by harishnelakurthi**
## Description
Add destination cluster a query is routed to in the cookie. More details here of the use-case here: https://github.com/trinodb/trino-gateway/issues/465

Fixes https://github.com/trinodb/trino-gateway/issues/465

## Release notes

```markdown
* The cookie data for requests to `/v1/statement` API would include the cluster to which a query got routed named as `trinoClusterHost`.
```

### CODE REVIEW

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/HaGatewayConfiguration.java [hunk 1]
```diff
@@ -97,6 +98,16 @@ public RoutingConfiguration getRouting()
         return routing;
     }
 
+    public void setIncludeClusterHostInResponse(boolean includeClusterHostInResponse)
+    {
+        this.includeClusterHostInResponse = includeClusterHostInResponse;
```
Comment (wendigo) [inline gateway-ha/src/main/java/io/trino/gateway/ha/config/HaGatewayConfiguration.java:103]: Config setters usually `return this` which allows for chaining

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/HaGatewayConfiguration.java [hunk 2]
```diff
@@ -97,6 +98,16 @@ public RoutingConfiguration getRouting()
...
+    }
+
+    public boolean getIncludeClusterHostInResponse()
+    {
+        return this.includeClusterHostInResponse;
```
Comment (wendigo) [inline gateway-ha/src/main/java/io/trino/gateway/ha/config/HaGatewayConfiguration.java:108]: drop `this.` - redundant

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/HaGatewayConfiguration.java [hunk 3]
```diff
@@ -97,6 +98,16 @@ public RoutingConfiguration getRouting()
         return routing;
     }
 
+    public void setIncludeClusterHostInResponse(boolean includeClusterHostInResponse)
+    {
+        this.includeClusterHostInResponse = includeClusterHostInResponse;
+    }
+
+    public boolean getIncludeClusterHostInResponse()
```
Comment (wendigo) [inline gateway-ha/src/main/java/io/trino/gateway/ha/config/HaGatewayConfiguration.java:106]: should be: isIncludeClusterHostInResponse

File: gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java [hunk 4]
```diff
@@ -173,9 +177,14 @@ private void performRequest(
             future = future.transform(response -> recordBackendForQueryId(request, response), executor);
         }
 
+        Map<String, String> extraHeaders = new HashMap<>();
```
Comment (wendigo) [inline gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:180]: The mutable maps should be avoided.

This could be just:
`Map<String, String> extraHeaders = includeClusterInfoInResponse ? ImmutableMap.of(CLUSTER_HOST_HEADER, remoteUri.getHost()) : ImmutableMap.of()`

File: gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java [hunk 5]
```diff
@@ -206,10 +215,15 @@ private static String getRemoteTarget(URI remoteUri)
         return format("%s://%s", remoteUri.getScheme(), remoteUri.getAuthority());
     }
 
-    private Response buildResponse(ProxyResponse response, ImmutableList<NewCookie> cookie)
+    private Response buildResponse(ProxyResponse response, ImmutableList<NewCookie> cookie, Map<String, String> extraHeaders)
     {
         Response.ResponseBuilder builder = Response.status(response.statusCode()).entity(response.body());
         response.headers().forEach((headerName, value) -> builder.header(headerName.toString(), value));
+
+        // Add any extra headers.
```
Comment (wendigo) [inline gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:223]: just: `extraHeaders.forEach(builder::header)`

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/HaGatewayConfiguration.java (hunk 1 above)
Comment (harishnelakurthi) [reply gateway-ha/src/main/java/io/trino/gateway/ha/config/HaGatewayConfiguration.java:103]: I see that all the other setters are returning void. Just wanted to be consistent. Let me know if we still want to change it.

**Comment by willmostly** [comment]
@harishnelakurthi please sign the CLA so that we can take this forward, and address all comments. Please squash your commits as well.

**Comment by harishnelakurthi** [comment]
@willmostly - I have signed the CLA and emailed a week ago. Also, I've addressed all the comments on the PR except there was an open question which I have responded to the reviewer. Let me know if anything is needed from my end.

**Comment by willmostly** [comment]
@harishnelakurthi I think you may have missed my comment above. I would like to avoid adding a new header just to pass back information unrelated to the routing protocol. This should be a cookie. You can also update the logic such that the cookie is only added/updated in the response to a POST to `v1/statement`. Please update the PR title and description to reflect the change as well.

**Comment by harishnelakurthi** [comment]
Thanks, @willmostly I have made the changes to pass this data via session cookie. Please review and let me know if any changes are needed. Also, updated the PR title and description. Will squash the commits before merge.

**Comment by andythsu** [comment]
Can we change the PR name to "cluster" instead of "backend"

**Comment by willmostly** [comment]
@harishnelakurthi you can use Airlift's codestyle in your IDE to avoid builds failing due to checkstyle https://github.com/airlift/codestyle

File: gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java [hunk 6]
```diff
@@ -169,13 +172,19 @@ private void performRequest(
...
+                // add trinoClusterHost to sessionCookies
+                sessionCookies.add(new NewCookie.Builder("trinoClusterHost").value(remoteUri.getHost()).build());
+            }
         }
 
+        ImmutableList<NewCookie> finalSessionCookie = ImmutableList.copyOf(sessionCookies);
         setupAsyncResponse(
                 asyncResponse,
-                future.transform(response -> buildResponse(response, oauth2GatewayCookie), executor)
+                future.transform(response -> buildResponse(response, oauth2GatewayCookie, finalSessionCookie), executor)
```
Comment (willmostly) [inline gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:187]: This is a bit more awkward than it needs to be because `getOAuth2GatewayCookie` returns an `ImmutableList`. Can you change it's return type to `List` and update the method to use some mutable implementation like `ArrayList`, then add your `trinoClusterHost` cookie to that list? Then you can revert the changes to `buildResponse`

File: gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java [hunk 7]
```diff
@@ -169,13 +172,19 @@ private void performRequest(
 
         FluentFuture<ProxyResponse> future = executeHttp(request);
 
+        List<NewCookie> sessionCookies = new ArrayList<>();
         if (statementPaths.stream().anyMatch(request.getUri().getPath()::startsWith) && request.getMethod().equals(HttpMethod.POST)) {
             future = future.transform(response -> recordBackendForQueryId(request, response), executor);
+            if (includeClusterInfoInResponse) {
+                // add trinoClusterHost to sessionCookies
```
Comment (willmostly) [inline gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:179]: not: comment here is unnecessary, the code is self-explanatory

**Review by willmostly — COMMENTED** [review]
Looks pretty close % comments. 

Can you add a test for this functionality? `TestGatewayHaMultipleBackend` should have the infrastructure you need

File: gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java [hunk 8]
```diff
@@ -171,34 +174,39 @@ private void performRequest(
 
         if (statementPaths.stream().anyMatch(request.getUri().getPath()::startsWith) && request.getMethod().equals(HttpMethod.POST)) {
             future = future.transform(response -> recordBackendForQueryId(request, response), executor);
+            if (includeClusterInfoInResponse) {
+                oauth2GatewayCookie.add(new NewCookie.Builder("trinoClusterHost").value(remoteUri.getHost()).build());
+            }
         }
 
+        ImmutableList<NewCookie> cookie = ImmutableList.copyOf(oauth2GatewayCookie);
```
Comment (willmostly) [inline gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:182]: alter `buildResponse` to accept a `List` instead of copying `oauth2GatewayCookie` to a new `ImmutableList`.

File: gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java [hunk 9]
```diff
@@ -160,7 +163,7 @@ private void performRequest(
             addXForwardedHeaders(servletRequest, requestBuilder);
         }
 
-        ImmutableList<NewCookie> oauth2GatewayCookie = getOAuth2GatewayCookie(remoteUri, servletRequest);
+        List<NewCookie> oauth2GatewayCookie = getOAuth2GatewayCookie(remoteUri, servletRequest);
```
Comment (willmostly) [inline gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:166]: rename this to `cookies` since we will add a non-oauth2 cookie to the list

File: gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java [hunk 10]
```diff
@@ -145,6 +145,11 @@ public void testQueryDeliveryToMultipleRoutingGroups()
                         .build();
         Response response1 = httpClient.newCall(request1).execute();
         assertThat(response1.body().string()).contains("http://localhost:" + routerPort);
+
+        List<Cookie> cookies = Cookie.parseAll(response1.request().url(), response1.headers());
+        Optional<Cookie> cookie = cookies.stream().filter(c -> c.name().equals("trinoClusterHost")).findAny();
+        assertThat(cookie).isNotEmpty();
```
Comment (willmostly) [inline gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:151]: We should test for a specific value to make sure the cluster host is set correctly. You can control which backend the request is sent to using the `X-Trino-Routing-Group` header. When you set 
```
.addHeader("X-Trino-Routing-Group", "scheduled")
```
then the backend should be `http://localhost:backend2Port`, when `X-Trino-Routing-Group` is unset you'll get `http://localhost:backend1Port`. You should also use the test to define the behavior when the initial client request already contains a cookie with the key `trinoClusterHost` - I expect the value should be overwritten. It will be cleaner to make a new test method.

**Review by willmostly — CHANGES_REQUESTED** [review]
Minor code comments and some additional testing

File: gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java [hunk 11]
```diff
@@ -145,6 +145,7 @@ public void testQueryDeliveryToMultipleRoutingGroups()
                         .build();
         Response response1 = httpClient.newCall(request1).execute();
         assertThat(response1.body().string()).contains("http://localhost:" + routerPort);
+
```
Comment (willmostly) [inline gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:148]: nit: remove this additional newline

**Review by willmostly — APPROVED** [review]
I think this is ready. Please remove the stray newline from the `testQueryDeliveryToMultipleRoutingGroups` method and we can get this merged

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/HaGatewayConfiguration.java [hunk 12]
```diff
@@ -97,6 +98,16 @@ public RoutingConfiguration getRouting()
         return routing;
     }
 
+    public void setIncludeClusterHostInResponse(boolean includeClusterHostInResponse)
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/config/HaGatewayConfiguration.java:101]: Please move this getter and setter to the bottom for consistency with field order.

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/HaGatewayConfiguration.java [hunk 13]
```diff
@@ -97,6 +98,16 @@ public RoutingConfiguration getRouting()
...
+    {
+        this.includeClusterHostInResponse = includeClusterHostInResponse;
+    }
+
+    public boolean isIncludeClusterHostInResponse()
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/config/HaGatewayConfiguration.java:106]: Replace the order of getter and setter.

File: gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java [hunk 14]
```diff
@@ -160,7 +162,7 @@ private void performRequest(
             addXForwardedHeaders(servletRequest, requestBuilder);
         }
 
-        ImmutableList<NewCookie> oauth2GatewayCookie = getOAuth2GatewayCookie(remoteUri, servletRequest);
+        List<NewCookie> cookies = getOAuth2GatewayCookie(remoteUri, servletRequest);
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:165]: Please use `ImmutableList.Builder` instead.

File: gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java [hunk 15]
```diff
@@ -171,46 +173,50 @@ private void performRequest(
...
                 asyncResponse,
-                future.transform(response -> buildResponse(response, oauth2GatewayCookie), executor)
+                future.transform(response -> buildResponse(response, cookies), executor)
                         .catching(ProxyException.class, e -> handleProxyException(request, e), directExecutor()));
     }
 
-    private ImmutableList<NewCookie> getOAuth2GatewayCookie(URI remoteUri, HttpServletRequest servletRequest)
+    private List<NewCookie> getOAuth2GatewayCookie(URI remoteUri, HttpServletRequest servletRequest)
     {
+        List<NewCookie> cookies = new ArrayList<>();
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:189]: Please use `ImmutableList.Builder` instead.

File: gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java [hunk 16]
```diff
@@ -158,6 +158,49 @@ public void testQueryDeliveryToMultipleRoutingGroups()
         assertThat(response4.body().string()).contains("http://localhost:" + routerPort);
     }
 
+    @Test
+    public void testTrinoClusterHostCookie()
+            throws Exception
+    {
+        RequestBody requestBody =
+                RequestBody.create(MediaType.parse("application/json; charset=utf-8"), "SELECT 1");
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:166]: Please avoid using deprecated `RequestBody.create` method. Also, pleas fix other waring in this test.

File: gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java [hunk 17]
```diff
@@ -158,6 +158,49 @@ public void testQueryDeliveryToMultipleRoutingGroups()
...
+
+        // When X-Trino-Routing-Group is set in header, query should be routed to cluster under the
+        // routing group
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:169]: nit: Unwrap.

File: gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java [hunk 18]
```diff
@@ -158,6 +158,49 @@ public void testQueryDeliveryToMultipleRoutingGroups()
...
+                new Request.Builder()
+                        .url("http://localhost:" + routerPort + "/v1/statement")
+                        .addHeader("X-Trino-User", "test")
+                        .post(requestBody)
+                        .addHeader("X-Trino-Routing-Group", "scheduled")
+                        .build();
+        Response response1 = httpClient.newCall(request1).execute();
+        assertThat(response1.body().string()).contains("http://localhost:" + routerPort);
+        List<Cookie> cookies = Cookie.parseAll(response1.request().url(), response1.headers());
+        Optional<Cookie> cookie = cookies.stream().filter(c -> c.name().equals("trinoClusterHost")).findAny();
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:180]: Please consider replacing `findAny()` with `collect(onlyElement())`.

File: gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java [hunk 19]
```diff
@@ -158,6 +158,49 @@ public void testQueryDeliveryToMultipleRoutingGroups()
...
+        assertThat(cookie).isNotEmpty();
+        String expectedHost = "localhost";
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:182]: Inline this variable.

File: gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java [hunk 20]
```diff
@@ -158,6 +158,49 @@ public void testQueryDeliveryToMultipleRoutingGroups()
...
+                        .post(requestBody)
+                        .addHeader("X-Trino-Routing-Group", "scheduled")
+                        .addHeader("Cookie", "trinoClientHost=foo.example.com")
+                        .build();
+        Response response2 = httpClient.newCall(request2).execute();
+        assertThat(response2.body().string()).contains("http://localhost:" + routerPort);
+        cookies = Cookie.parseAll(response1.request().url(), response1.headers());
+        cookie = cookies.stream().filter(c -> c.name().equals("trinoClusterHost")).findAny();
+        assertThat(cookie).isNotEmpty();
+        expectedHost = "localhost";
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:200]: Inline this variable.

File: gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java [hunk 21]
```diff
@@ -158,6 +158,49 @@ public void testQueryDeliveryToMultipleRoutingGroups()
...
+        assertThat(response1.body().string()).contains("http://localhost:" + routerPort);
+        List<Cookie> cookies = Cookie.parseAll(response1.request().url(), response1.headers());
+        Optional<Cookie> cookie = cookies.stream().filter(c -> c.name().equals("trinoClusterHost")).findAny();
+        assertThat(cookie).isNotEmpty();
+        String expectedHost = "localhost";
+        assertThat(cookie.orElseThrow().value()).isEqualTo(expectedHost);
+        // test with sending the request which includes trinoClusterHost in the cookie
+        // When X-Trino-Routing-Group is set in header, query should be routed to cluster under the
+        // routing group
+        Request request2 =
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:187]: Please avoid number suffix as much as possible.

**Review by ebyhr — CHANGES_REQUESTED** [review]

**Comment by ebyhr** [comment]
Please squash commits into one.

File: gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java (hunk 14 above)
Comment (harishnelakurthi) [reply gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:165]: This is intentionally a mutable list. Please check the code later which modifies the list for adding more cookies.

File: gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java (hunk 15 above)
Comment (harishnelakurthi) [reply gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:189]: Refer to my other comment.

File: gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java (hunk 16 above)
Comment (harishnelakurthi) [reply gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:166]: All the existing tests use the same. Can you please provide the right method to use here?

File: gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java (hunk 14 above)
Comment (ebyhr) [reply gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:165]: I'm suggesting `ImmutableList.Builder` not `ImmutableList`

File: gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java (hunk 15 above)
Comment (ebyhr) [reply gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:189]: I'm suggesting `ImmutableList.Builder` not `ImmutableList`

Comment (ebyhr) [reply gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:189]: Actually, changes in this method aren't necessary. Please revert.

Comment (harishnelakurthi) [reply gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:189]: Please suggest alternative instead.

File: gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java [hunk 22]
```diff
@@ -160,7 +162,9 @@ private void performRequest(
             addXForwardedHeaders(servletRequest, requestBuilder);
         }
 
-        ImmutableList<NewCookie> oauth2GatewayCookie = getOAuth2GatewayCookie(remoteUri, servletRequest);
+        ImmutableList.Builder<NewCookie> cookieBuilder = new ImmutableList.Builder<>();
+        ImmutableList<NewCookie> oauthCookies = getOAuth2GatewayCookie(remoteUri, servletRequest);
+        oauthCookies.forEach(cookieBuilder::add);
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:167]:
```java
        ImmutableList.Builder<NewCookie> cookieBuilder = ImmutableList.builder();
        cookieBuilder.addAll(getOAuth2GatewayCookie(remoteUri, servletRequest));
```

File: gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java [hunk 23]
```diff
@@ -158,6 +159,43 @@ public void testQueryDeliveryToMultipleRoutingGroups()
...
+                new Request.Builder()
+                        .url("http://localhost:" + routerPort + "/v1/statement")
+                        .addHeader("X-Trino-User", "test")
+                        .post(requestBody)
+                        .addHeader("X-Trino-Routing-Group", "scheduled")
+                        .build();
+        Response response1 = httpClient.newCall(request1).execute();
+        assertThat(response1.body().string()).contains("http://localhost:" + routerPort);
+        List<Cookie> cookies = Cookie.parseAll(response1.request().url(), response1.headers());
+        Cookie cookie = cookies.stream().filter(c -> c.name().equals("trinoClusterHost")).collect(MoreCollectors.onlyElement());
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:180]: Static import `onlyElement()`.

**Comment by harishnelakurthi** [comment]
@ebyhr @willmostly Addressed all the comments.If there are no comments I can squash all the commits and would appreciate merging this in as it's been hanging around for a while.

File: gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java (hunk 22 above)
Comment (ebyhr) [reply gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java]: > ImmutableList.Builder<NewCookie> cookieBuilder = ImmutableList.builder();

This isn't addressed yet.

**Comment by ebyhr** [comment]
Please remove redundant commit body.

File: gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java (hunk 22 above)
Comment (harishnelakurthi) [reply gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java]: DOne

**Comment by harishnelakurthi** [comment]
@ebyhr - I've addressed all the comments on this. Can this be merged please?

File: gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java [hunk 24]
```diff
@@ -158,6 +159,42 @@ public void testQueryDeliveryToMultipleRoutingGroups()
         assertThat(response4.body().string()).contains("http://localhost:" + routerPort);
     }
 
+    @Test
+    public void testTrinoClusterHostCookie()
+            throws Exception
+    {
+        RequestBody requestBody = RequestBody.create("SELECT 1", MediaType.get("application/json; charset=utf-8"));
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:166]: Please rebase and use `MEDIA_TYPE` constant.

File: gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java [hunk 25]
```diff
@@ -158,6 +159,42 @@ public void testQueryDeliveryToMultipleRoutingGroups()
...
+
+        // When X-Trino-Routing-Group is set in header, query should be routed to cluster under the routing group
+        Request requestWithoutCookie =
+                new Request.Builder()
+                        .url("http://localhost:" + routerPort + "/v1/statement")
+                        .addHeader("X-Trino-User", "test")
+                        .post(requestBody)
+                        .addHeader("X-Trino-Routing-Group", "scheduled")
+                        .build();
+        Response responseOne = httpClient.newCall(requestWithoutCookie).execute();
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:176]: Rename `responseOne` and `responseTwo` variables.

File: gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java [hunk 26]
```diff
@@ -158,6 +159,42 @@ public void testQueryDeliveryToMultipleRoutingGroups()
...
+        assertThat(responseOne.body().string()).contains("http://localhost:" + routerPort);
+        List<Cookie> cookies = Cookie.parseAll(responseOne.request().url(), responseOne.headers());
+        Cookie cookie = cookies.stream().filter(c -> c.name().equals("trinoClusterHost")).collect(onlyElement());
+        assertThat(cookie.value()).isEqualTo("localhost");
+        // test with sending the request which includes trinoClusterHost in the cookie
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:181]: nit: Add an empty line.

File: gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java [hunk 27]
```diff
@@ -158,6 +159,42 @@ public void testQueryDeliveryToMultipleRoutingGroups()
...
+                        .url("http://localhost:" + routerPort + "/v1/statement")
+                        .addHeader("X-Trino-User", "test")
+                        .post(requestBody)
+                        .addHeader("X-Trino-Routing-Group", "scheduled")
+                        .addHeader("Cookie", "trinoClientHost=foo.example.com")
+                        .build();
+        Response responseTwo = httpClient.newCall(requestWithCookie).execute();
+        assertThat(responseTwo.body().string()).contains("http://localhost:" + routerPort);
+        cookies = Cookie.parseAll(responseTwo.request().url(), responseTwo.headers());
+        cookie = cookies.stream().filter(c -> c.name().equals("trinoClusterHost")).collect(onlyElement());
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:194]: Please avoid reusing `cookies` variable.

File: gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java [hunk 28]
```diff
@@ -158,6 +159,42 @@ public void testQueryDeliveryToMultipleRoutingGroups()
...
+        assertThat(cookie.value()).isEqualTo("localhost");
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:195]: > When X-Trino-Routing-Group is set in header, query should be routed to cluster under the routing group
> ...
> test with sending the request which includes trinoClusterHost in the cookie when X-Trino-Routing-Group is set in header, query should be routed to cluster under the routing group

Values in assertions between `requestWithoutCookie` and `requestWithCookie` are the same. I don't understand what you want to express in this comment.

**Review by ebyhr — COMMENTED** [review]
Looks good except for comments.

File: gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java (hunk 28 above)
Comment (harishnelakurthi) [reply gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:195]: I'll fix the comment. The intent of this test is to make sure that the cookie is overriden correctly even though the request sends with some random value.

**Review by ebyhr — APPROVED** [review]

**Comment by ebyhr** [comment]
@harishnelakurthi Please squash commits into one.

**Comment by harishnelakurthi** [comment]
Thanks a lot @ebyhr for all the comments. I've squashed all the commits. Can you merge the PR if it's all good? Let me know if anything else needs to be done. Thanks!

**Comment by ebyhr** [comment]
@harishnelakurthi Could you please use "rebase" instead of "merge"? 
Screenshot 2024-09-27 at 6 55 52
